src/Microsoft.Tye.Hosting/HttpProxyService.cs

```diff
@@ -117,23 +122,30 @@ public async Task StartAsync(Application application)
                         {
                             var port = targetBinding.ReplicaPorts[i];
                             var url = $"{targetBinding.Protocol}://localhost:{port}";
-                            uris.Add(new Uri(url));
+                            uris.Add((port, new Uri(url)));
                         }
 
                         _logger.LogInformation("Service {ServiceName} is using {Urls}", targetServiceDescription.Name, string.Join(",", uris.Select(u => u.ToString())));
 
                         // The only load balancing strategy here is round robin
                         long count = 0;
-                        RequestDelegate del = context =>
+                        RequestDelegate del = async context =>
                         {
                             var next = (int)(Interlocked.Increment(ref count) % uris.Count);
+                            for (int i = 0; !_readyPorts.ContainsKey(uris[next].Port) && i < uris.Count; i++, next = (int)(Interlocked.Increment(ref count) % uris.Count)) ;
```

Comment:
Lot of stuff going on in this for loop, can you break it up a bit and add a comment about what it is doing?

My best guess is loop through all ready ports until we find one that is ready.

Reply:
that's right. I'm looping through all the ports (starting from the current selected port, `# of ports` times) to find one that is ready. I'll try to simplify and add comment